libr/bin/p/bin_dex.c

```diff
@@ -271,48 +273,34 @@ static char *dex_get_proto(RBinDexObj *bin, int proto_id) {
 	if (!r_buf_read_at (bin->b, params_off, params_buf, sizeof (params_buf))) {
 		return NULL;
 	}
-	// size of the list, in entries
-	list_size = r_read_le32 (params_buf);
-	if (list_size * sizeof (ut16) >= bin->size) {
-		return NULL;
-	}
-
-	for (i = 0; i < list_size; i++) {
-		int buff_len = 0;
-		int off = params_off + 4 + (i * 2);
-		if (off >= bin->size) {
-			break;
-		}
-		ut8 typeidx_buf[sizeof (ut16)];
-		if (!r_buf_read_at (bin->b, off, typeidx_buf, sizeof (typeidx_buf))) {
-			break;
-		}
-		type_idx = r_read_le16 (typeidx_buf);
-		if (type_idx >= bin->header.types_size || type_idx >= bin->size) {
-			break;
-		}
-		const char *buff = getstr (bin, bin->types[type_idx].descriptor_id);
-		if (!buff) {
-			break;
-		}
-		buff_len = strlen (buff);
-		size += buff_len + 1;
-		char *newsig = realloc (signature, size);
-		if (!newsig) {
-			eprintf ("Cannot realloc to %d\n", size);
-			free (signature);
-			break;
+	// size of the list, in 16 bit entries
+	ut32 list_size = r_read_le32 (params_buf);
+	size_t typeidx_bufsize = (list_size * sizeof (ut16));
```

Comment:
no need for parentheses

Reply:
why `size_t`? list_size is ut32 and `params_off`, used in the next line, is `ut32` as well. Maybe we could use `ut64` if we want to be sure the result of the multiplication does not overflow.

Reply:
ut32 * 2 can result in integer overflow. Also typeidx_bufsize is later used to specify the size of the allocation, on 64bit systems size_t is 64bit wide, on 32bit it may fail.

One solution to this is to check for the overflow or clamp the list_size to avoid overflowing the 32bit boundary. Also, i doubt any dex is that large to hold 4 million types when the method number limit is 65,536. I may check dvm's implementation to be more precise, but i don't think it actually matters that much on real life use cases.

So options are:

* use ut64
* avoid the overflow and chop the size or fail early

by checking the google implementation:

```
128         int returnIdx = file.getTypeIds().indexOf(prototype.getReturnType());
129         int paramsOff = OffsettedItem.getAbsoluteOffsetOr0(parameterTypes);
```

it's using `int` 🤦  so I would say the trimming with warning is the best option here. But i'm already checking against file->size. So that may be enough probably.